python/task/workflow.py

```diff
@@ -0,0 +1,703 @@
+    used for selecting an appropriate reader task for the format."""
+    if isinstance(input, list):
+      outputs = []
+      shards = len(input)
+      for shard in xrange(shards):
+        format = input[shard].format
+        if type(format) == str: format = Format(format)
+        if format == None: format = Format("text")
+        tasktype = readers.get(format.file)
+        if tasktype == None: raise Exception("No reader for " + str(format))
+
+        reader = self.task(tasktype, name=name, shard=Shard(shard, shards))
+        reader.attach_input("input", input[shard])
+        output = self.channel(reader, format=format.as_message())
+        outputs.append(output)
+      return outputs
+    else:
+      format = input.format
+      if type(format) == str: format = Format(format)
+      if format == None: format = Format("text")
+      tasktype = readers.get(format.file)
+      if tasktype == None: raise Exception("No reader for " + str(format))
+
+      reader = self.task(tasktype, name=name)
+      reader.attach_input("input", input)
+      output = self.channel(reader, format=format.as_message())
+      return output
+
+  def write(self, producer, output, sharding=None, name=None):
+    """Add writers for output resource(s). The format of the output resource is
+    used for selecting an appropriate reader task for the format."""
+    # Determine fan-in (channels) and fan-out (files).
+    if not isinstance(producer, list): producer = [producer]
+    if not isinstance(output, list): output = [output]
+    fanin = len(producer)
+    fanout = len(output)
+
+    # Use sharding if fan-out is different from fan-in.
+    if sharding == None and fanout != 1 and fanin != fanout:
+      sharding = "sharder"
+
+    # Create sharder if needed.
+    if sharding == None:
+      input = producer
+    else:
+      sharder = self.task(sharding)
+      if fanin == 1:
+        self.connect(producer[0], sharder)
+      else:
+        self.connect(producer, sharder)
+      input = self.channel(sharder, shards=fanout, format=producer[0].format)
+
+    # Create writer tasks for writing to output.
+    writer_tasks = []
+    for shard in xrange(fanout):
+      format = output[shard].format
+      if type(format) == str: format = Format(format)
+      if format == None: format = Format("text")
+      tasktype = writers.get(format.file)
+      if tasktype == None: raise Exception("No writer for " + str(format))
+
+      if fanout == 1:
+        writer = self.task(tasktype, name=name)
+      else:
+        writer = self.task(tasktype, name=name, shard=Shard(shard, fanout))
+      writer.attach_output("output", output[shard])
+      writer_tasks.append(writer)
+
+    # Connect producer(s) to writer task(s).
+    if isinstance(input, list) and len(input) == 1: input = input[0]
+    if fanout == 1: writer_tasks = writer_tasks[0]
+    self.connect(input, writer_tasks)
+
+    return output
```

Comment:
did you intend to return writer_tasks here? 'output' is already an input of this method.

Reply:
I think this is what I intended. I agree that returning one of the input arguments is not really providing any new information to the caller, but I think it is more convenient for the caller to get the list of output resource rather than the list of writer tasks, although this remains to be seen since we are not really using this anywhere.